src/symmetric_crypto.cpp

```diff
@@ -0,0 +1,382 @@
+        assert(EVPCipherConstructor && "ECP cipher constructor must be available here.");
+
+        openssl::_EVP_CipherInit_ex(_ctx.get(),
+                                    EVPCipherConstructor(),
+                                    nullptr,
+                                    nullptr,
+                                    nullptr,
+                                    this->_operation == Operation::Encryption);
+
+        // Check id key size matches cipher block size
+        size_t expectedKeySize = openssl::_EVP_CIPHER_CTX_key_length(_ctx.get());
+        if (secretKey.size() != expectedKeySize) {
+            auto formatter = boost::format("Invalid size of Key %d bytes. Must be %d bytes.");
+            formatter % iv.size() % expectedKeySize;
+            throw MoCOCrWException(formatter.str());
+        }
+
+        // Check IV length and adjust if cipher supports it
+        switch (mode) {
+            case SymmetricCipherMode::GCM: {
+                openssl::_EVP_CIPHER_CTX_ctrl(_ctx.get(), EVP_CTRL_GCM_SET_IVLEN, iv.size(),
+                                              nullptr);
+            } break;
+            case SymmetricCipherMode::CTR:
+                //[[fallthrough]];
+            case SymmetricCipherMode::CBC: {
+                size_t expectedIVSize = openssl::_EVP_CIPHER_CTX_iv_length(_ctx.get());
+                if (iv.size() < expectedIVSize) {
+                    auto formatter = boost::format(
+                            "Invalid size of IV %d bytes. Must be %d bytes.");
+                    formatter % iv.size() % expectedIVSize;
+                    throw MoCOCrWException(formatter.str());
+                }
+            } break;
+        }
+
+        openssl::_EVP_CipherInit_ex(_ctx.get(),
+                                    EVPCipherConstructor(),
+                                    nullptr,
+                                    secretKey.data(),
+                                    iv.data(),
+                                    this->_operation == Operation::Encryption);
+
+        switch (padding) {
+            case SymmetricCipherPadding::PKCS:
+                // OpenSSL uses PKCS padding by default.
+                break;
+        }
+    };
+
+    void update(const std::vector<uint8_t> &plaintext)
+    {
+        if (_isFinished) {
+            throw MoCOCrWException(
+                    "Further calls to update() are not allowed once finish() was called.");
+        }
+
+        int processingChunkSize = plaintext.size();
+        std::vector<uint8_t> processedChunk(processingChunkSize + EVP_MAX_BLOCK_LENGTH);
+
+        openssl::_EVP_CipherUpdate(_ctx.get(),
+                                   processedChunk.data(),
+                                   &processingChunkSize,
+                                   plaintext.data(),
+                                   plaintext.size());
+
+        processedChunk.resize(processingChunkSize);
+        _bufferStrategy->write(processedChunk);
```

Comment:
Shouldn't we move `processedChunk` into `_bufferStrategy->write()` in order to avoid copying the data and change the buffer strategy to accept `std::vector<uint8_t>&&` only?

Reply:
good point